### PR TITLE
Feature: Add conditional timelocking to multisig wallet

### PR DESCRIPTION
**Pull request by nadavhollander**
This PR introduces the following changes:

 - Adds a parametrizable timelock to the `MultiSigWallet` that prevents a confirmed transaction from being executed until a pre-configured timelock period passes.
 - Patches migrations to use new `DharmaMultiSigWallet`

Tests will likely fail on this until all branches where test patching is implemented are merged into this.

### CODE REVIEW

File: contracts/wallet/DharmaMultiSigWallet.sol [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+            txn.executed = false;
+        }
+    }
+
+    /**
+     * Internal function that records the time at which
+     * a transaction reached the minimal threshold of
+     * confirmations.
+     */
+    function setConfirmationTime(uint transactionId, uint confirmationTime)
```
Comment (graemecode) [inline contracts/wallet/DharmaMultiSigWallet.sol]: `_ setConfirmationTime`

File: contracts/wallet/DharmaMultiSigWallet.sol [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+        ConfirmationTimeSet(transactionId, confirmationTime);
+    }
+
+    /**
+     * Returns true if the first four bytes of the given data matches
+     * the function selector Method ID for the "pause" method --
+     * for more information, see the Ethereum wiki's contract ABI specification:
+     * https://github.com/ethereum/wiki/wiki/Ethereum-Contract-ABI
+     */
+    function isPauseFunctionTransaction(bytes data)
```
Comment (graemecode) [inline contracts/wallet/DharmaMultiSigWallet.sol]: `_ isPauseFunctionTransaction`

File: contracts/wallet/BaseMultiSigWallet.sol [hunk 3]
```diff
@@ -3,7 +3,7 @@ pragma solidity 0.4.18;
 
 /// @title Multisignature wallet - Allows multiple parties to agree on transactions before execution.
 /// @author Stefan George - <stefan.george@consensys.net>
-contract MultiSigWallet {
+contract BaseMultiSigWallet {
```
Comment (saturnial) [inline contracts/wallet/BaseMultiSigWallet.sol]: Should we make this abstract? Or a library -- basically anything that prevents it from being instantiable on its own.

File: contracts/wallet/DharmaMultiSigWallet.sol [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+
+    /**
+     * Changes the duration of the time lock for transactions.  Can only be called by the wallet.
+     */
+    function changeTimeLock(uint _timelockInSeconds)
+        public
+        onlyWallet
+    {
+        timelockInSeconds = _timelockInSeconds;
+        TimeLockChange(_timelockInSeconds);
```
Comment (saturnial) [inline contracts/wallet/DharmaMultiSigWallet.sol]: Where is this event specified?

File: contracts/wallet/DharmaMultiSigWallet.sol [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+    {
+        confirmations[transactionId][msg.sender] = false;
+        Revocation(msg.sender, transactionId);
+    }
+
+    /**
+     * Allows anyone to execute a transaction if it has been
+     * confirmed and the timelock has passed.
+     */
+    function executeTransaction(uint transactionId)
```
Comment (saturnial) [inline contracts/wallet/DharmaMultiSigWallet.sol]: Why anyone? Shouldn't they be an owner.

File: contracts/wallet/DharmaMultiSigWallet.sol [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+    {
+        _executeTransaction(transactionId);
+    }
+
+    /**
+     * In case of an emergency, allows anyone to execute a transaction if it has
+     * been confirmed and matches the "pause" function signature WITHOUT waiting
+     * for the timelock to pass.
+     */
+    function executePauseTransactionImmediately(uint transactionId)
```
Comment (saturnial) [inline contracts/wallet/DharmaMultiSigWallet.sol]: Why is this possible?

File: contracts/wallet/BaseMultiSigWallet.sol (hunk 3 above)
Comment (saturnial) [reply contracts/wallet/BaseMultiSigWallet.sol]: Let's table this for now.

File: contracts/wallet/DharmaMultiSigWallet.sol (hunk 4 above)
Comment (saturnial) [reply contracts/wallet/DharmaMultiSigWallet.sol]: This is specified in a later PR.

File: contracts/wallet/DharmaMultiSigWallet.sol (hunk 5 above)
Comment (saturnial) [reply contracts/wallet/DharmaMultiSigWallet.sol]: As discussed IRL, this is fine.

**Pull request merged**